### DISCUSSION TITLE
Should we cancel sync to avoid retries when the device is largely offline?

### DISCUSSION THREAD

**Discussion by ellykits** [opensrp/fhircore#2021] [Q&A] [answered]
Currently we schedule periodic sync when the application is started either in online or offline mode. When the device is offline does it make sense to still schedule the sync? If we don't we at some point need to run sync when the device gets back online so we will need some sort of listener to know when the device is back online. The sync failure exception is currently handled so there would be no app crashes even if the device attempt to sync data while offline.

**Comment by ndegwamartin** [comment, marked as answer]
_Workmanager_ provides a `Constraints` class object which we could use to specify that certain jobs should only run if there's an internet connection 

**For example:**

```
Constraints constraints = new Constraints.Builder()
    .setRequiredNetworkType(NetworkType.CONNECTED)
    .build();

PeriodicWorkRequest request =
    new PeriodicWorkRequest.Builder(MyWorker.class, 15, TimeUnit.MINUTES)
        // Sets the input data for the ListenableWorker
        .setConstraints(constraints)
        .build();
```

We could also use `NetworkType.NOT_REQUIRED` to signal that that specific job **Does Not** require connectivity 

For more see the documentation here: [Android Background Work](https://developer.android.com/guide/background/persistent/getting-started/define-work)

**Reply by ellykits** [reply]
Marked as answer because with this configuration we do not get sync failure notifications anymore when device is offline.

**Comment by ndegwamartin** [comment]
Not sure we have this configured/factored in but the underlying _JobScheduler API_ in Android limits the Workers to prevent them from running for more than 10 minutes. After that interval the OS stops the Worker. 

There are recommended workarounds for supporting such long running tasks documented here. https://developer.android.com/guide/background/persistent/how-to/long-running

cc @dubdabasoduba @f-odhiambo

**Reply by ellykits** [reply]
Default configuration is to attempt retries 3 times (when a job fails) linearly with a backoff delay of 30 seconds. We can configure this if desired.
